src/functions.php

```diff
@@ -12,22 +12,29 @@
  * Creates a `Promise` which resolves with the stream data buffer
  *
  * @param ReadableStreamInterface $stream
+ * @param int $maxLength
  * @return Promise\CancellablePromiseInterface Promise<string, Exception>
  */
-function buffer(ReadableStreamInterface $stream)
+function buffer(ReadableStreamInterface $stream, $maxLength = 0)
 {
     // stream already ended => resolve with empty buffer
     if (!$stream->isReadable()) {
         return Promise\resolve('');
     }
 
+    $deferred = new Promise\Deferred();
     $buffer = '';
-    $bufferer = function ($data) use (&$buffer) {
+    $bufferer = function ($data) use (&$buffer, $deferred, $maxLength) {
         $buffer .= $data;
+        if ($maxLength > 0 && strlen($buffer) > $maxLength) {
+            $deferred->reject(new \RuntimeException('Buffer exceeded maximum length'));
+        }
```

Comment:
Should probably also stop buffering in the background?

Reply:
Actually when the promise is rejected it is automatically removing the buffer listener